Library/Paginator.swift

```diff
@@ -0,0 +1,136 @@
+
+  You can observe the results of `values`, `isLoading`, `error` and `state` to access the loaded data.
+
+ */
+
+public class Paginator<Envelope, Value: Equatable, Cursor, SomeError: Error, RequestParams> {
+  public enum Results: Equatable {
+    case unloaded
+    case someLoaded
+    case allLoaded
+    case empty
+    case error
+  }
+
+  @Published public var values: [Value]
+  @Published public var isLoading: Bool
+  @Published public var error: SomeError?
+  @Published public var state: Results
+
+  private var valuesFromEnvelope: (Envelope) -> [Value]
+  private var cursorFromEnvelope: (Envelope) -> Cursor?
+  private var requestFromParams: (RequestParams) -> AnyPublisher<Envelope, SomeError>
+  private var requestFromCursor: (Cursor) -> AnyPublisher<Envelope, SomeError>
+  private var cancellables = Set<AnyCancellable>()
+
+  private var lastCursor: Cursor?
+
+  public init(valuesFromEnvelope: @escaping ((Envelope) -> [Value]),
+              cursorFromEnvelope: @escaping ((Envelope) -> Cursor?),
+              requestFromParams: @escaping ((RequestParams) -> AnyPublisher<Envelope, SomeError>),
+              requestFromCursor: @escaping ((Cursor) -> AnyPublisher<Envelope, SomeError>)) {
+    self.values = []
+    self.isLoading = false
+    self.error = nil
+    self.state = .unloaded
+
+    self.valuesFromEnvelope = valuesFromEnvelope
+    self.cursorFromEnvelope = cursorFromEnvelope
+    self.requestFromParams = requestFromParams
+    self.requestFromCursor = requestFromCursor
+  }
+
+  func handleRequest(_ request: AnyPublisher<Envelope, SomeError>) {
+    request
+      .receive(on: RunLoop.main)
+      .catch { [weak self] error -> AnyPublisher<Envelope, SomeError> in
+        self?.error = error
+        self?.state = .error
+        return Empty<Envelope, SomeError>().eraseToAnyPublisher()
+      }
+      .assertNoFailure()
+      .handleEvents(receiveCompletion: { [weak self] _ in
+        self?.isLoading = false
+      }, receiveCancel: { [weak self] in
+        self?.isLoading = false
+      })
+      .sink(receiveValue: { [weak self] envelope in
+        guard let self = self else {
+          return
+        }
+
+        let newValues = self.valuesFromEnvelope(envelope)
+        self.values.append(contentsOf: newValues)
+
+        let cursor = self.cursorFromEnvelope(envelope)
+        self.lastCursor = cursor
+
+        if self.values.count == 0 {
+          self.state = .empty
+        } else if cursor == nil || newValues.count == 0 {
+          self.state = .allLoaded
+        } else {
+          self.state = .someLoaded
+        }
+      })
+      .store(in: &self.cancellables)
+  }
+
+  public func requestFirstPage(withParams params: RequestParams) {
+    self.cancel()
+
+    self.values = []
+    self.isLoading = true
+    self.error = nil
+
+    let request = self.requestFromParams(params)
+    self.handleRequest(request)
+  }
+
+  public func requestNextPage() {
```

Comment:
I find this kind of declarative implementation much easier to grok than the original functional implementation of `paginate`.